File: Python/Linked List/32_segregating_even_and_odd_nodes.py
"""
Idea:
keeping 2 pointers , each for odd and even numbers, and appending odd after even ones and returning
"""



class Solution:
    def divide(self, N, head):
        # Initializing odd and even linked lists
        odd = node()
        even = node()
        ptrODD = odd
        ptrEVEN = even
        temp = head

        # Segregating even and odd nodes
        while temp:
            if temp.data % 2 == 1:  # If node is odd
                ptrODD.next = temp
                ptrODD = ptrODD.next
            else:  # If node is even
                ptrEVEN.next = temp
                ptrEVEN = ptrEVEN.next
            temp = temp.next

        # Setting the last node of the odd list to None
        ptrODD.next = None

        # Combining the even and odd lists and returning the head of the even list
        ptrEVEN.next = odd.next
        return even.next


# code here


# {
# Driver Code Starts
# Initial Template for Python3

# Node Class
class node:
    def __init__(self):
        self.data = None
        self.next = None


# Linked List Class
class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, data):
        if self.head == None:
            self.head = node()
            self.tail = self.head
            self.head.data = data
        else:
            new_node = node()
            new_node.data = data
            new_node.next = None
            self.tail.next = new_node
            self.tail = self.tail.next


def printlist(head):
    temp = head
    while temp is not None:
        print(temp.data, end=" ")
        temp = temp.next
    print('')


# Driver Program
if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        list1 = Linked_List()
        n = int(input())
        values = list(map(int, input().strip().split()))
        for i in values:
            list1.insert(i)
        ob = Solution()
        newhead = ob.divide(n, list1.head)
        printlist(newhead)

# } Driver Code Ends

File: C++/Linked List/32_segregating_even_and_odd_nodes.cpp
/*
Idea:
keeping 2 pointers , each for odd and even numbers, and appending odd after even ones and returning
*/

//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};
void printList(Node* node)
{
    while (node != NULL) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << "\n";
}


// } Driver Code Ends
//User function template for C++

class Solution {
public:
    Node* divide(int N, Node* head)
    {
        // code here
        Node* odd = new Node(-1), * even = new Node(-1), * ptrODD = odd, * ptrEVEN = even, * temp = head;
        while (temp)
        {
            if (temp->data & 1)//if node is odd
            {
                ptrODD->next = temp;
                ptrODD = ptrODD->next;
            }
            else //if node is even
            {
                ptrEVEN->next = temp;
                ptrEVEN = ptrEVEN->next;
            }
            temp = temp->next;
        }
        //set the last node of the odd list to NULL
        ptrODD->next = NULL;

        // Combine the even and odd lists and return the head of the even list
        ptrEVEN->next = odd->next;
        return even->next;

    }
};

//{ Driver Code Starts.

int main() {
    //code
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin >> N;
        int data;
        cin >> data;
        struct Node* head = new Node(data);
        struct Node* tail = head;
        for (int i = 0; i < N - 1; ++i)
        {
            cin >> data;
            tail->next = new Node(data);
            tail = tail->next;
        }

        Solution ob;
        Node* ans = ob.divide(N, head);
        printList(ans);
    }
    return 0;
}

// } Driver Code Ends